### PR TITLE
Hashmod - Allow transactions to have attributes from the CSV

### PR DESCRIPTION
**Pull request by CrucisGamma**
### Description of the Change

Allow transaction to inherit attributes from a node.  Also significantly speeds up ingest of data from a CSV file

### Alternate Designs

Nil considered

### Why Should This Be In Core?

Already in core

### Benefits

Ingesting csv files is a lot quicker with the hashmod ingester now as it doesn't have to traverse the entire key chain each time.
Allows transactions to have an attribute from a vertex added to it.

### Possible Drawbacks

Nil known.

### Verification Process

Integration testing was done for new and larger csv files (one with 48K rows - when this same one took minutes previously, it took seconds to ingest with the changes)

### Applicable Issues

#1363

### CODE REVIEW

File: CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/Hashmod.java [hunk 1]
```diff
@@ -168,7 +168,16 @@ public String getFirstColumnOfTransaction(int transactionCol) {
     }
 
     public String getSecondColumnOfTransaction(int transactionCol) {
-        return getColumnOfTransaction(transactionCol, ".*?\\.\\.\\.([^\"]+)");
+        return getColumnOfTransaction(transactionCol, ".*?\\.\\.\\.([^\"]+?)(;;;.*|$)");
+    }
+
+    public String getTransactionAttribute(String attributeFromCSV) {
+        final Pattern transactionPattern = Pattern.compile("^.*;;;([^\"]+)");
+        Matcher matchPattern = transactionPattern.matcher(attributeFromCSV);
```
Comment (antares1470) [inline CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/Hashmod.java:176]: final

File: CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java [hunk 2]
```diff
@@ -120,14 +126,25 @@ private static void run(final GraphWriteMethods wg, final PluginInteraction inte
...
+                    final int newAttribute = wg.addAttribute(GraphElementType.VERTEX, newAttributeType, nextAttr, nextAttr, "", null);
+                    if (newAttribute != Graph.NOT_FOUND) {
+                        attributeValues[attrCount] = newAttribute;
+                        csvValues[attrCount] = i;
+                        attrCount++;
+                    }
+                }
+
+                if (createTransactions && !hashmod.getTransactionAttribute(nextAttr).equalsIgnoreCase("")) {
+                    String transactionAttributeName = hashmod.getTransactionAttribute(nextAttr);
```
Comment (antares1470) [inline CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java:139]: final

File: CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java [hunk 3]
```diff
@@ -120,14 +126,25 @@ private static void run(final GraphWriteMethods wg, final PluginInteraction inte
...
+                if (createAttributes) {
+                    final int newAttribute = wg.addAttribute(GraphElementType.VERTEX, newAttributeType, nextAttr, nextAttr, "", null);
+                    if (newAttribute != Graph.NOT_FOUND) {
+                        attributeValues[attrCount] = newAttribute;
+                        csvValues[attrCount] = i;
+                        attrCount++;
+                    }
+                }
+
+                if (createTransactions && !hashmod.getTransactionAttribute(nextAttr).equalsIgnoreCase("")) {
```
Comment (antares1470) [inline CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java:138]: I would change `!hashmod.getTransactionAttribute(nextAttr).equalsIgnoreCase("")` to `StringUtils.isNotBlank(hashmod.getTransactionAttribute(nextAttr))`

File: CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java [hunk 4]
```diff
@@ -176,17 +193,15 @@ private static void run(final GraphWriteMethods wg, final PluginInteraction inte
             interaction.notify(PluginNotificationLevel.WARNING, "Successfully updated " + numberSuccessful + "/" + vxPos + " nodes");
         }
 
-        if (createAllKeys) {
+        if (createVertices) {
             numberSuccessful = 0;
             for (final Entry<String, Integer> entry : keys.entrySet()) {
-                if (entry.getValue() == 0) {
-                    int newVertex = wg.addVertex();
+                int newVertexId = wg.addVertex();
```
Comment (antares1470) [inline CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java:199]: final

File: CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java [hunk 5]
```diff
@@ -233,7 +248,15 @@ private static void run(final GraphWriteMethods wg, final PluginInteraction inte
 
                             String attr2Value = wg.getObjectValue(transaction2Attribute, vx2Id);
                             if (attr1Value.equals(attr2Value) && vxId != vx2Id) {
-                                wg.addTransaction(vxId, vx2Id, false);
+                                int newTransactionId = wg.addTransaction(vxId, vx2Id, false);
```
Comment (antares1470) [inline CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java:251]: final

File: CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java [hunk 6]
```diff
@@ -233,7 +248,15 @@ private static void run(final GraphWriteMethods wg, final PluginInteraction inte
...
+
+
+                                for (i = 0; i < transAttrCount; i++) {
+                                    String theVal = wg.getStringValue(fromNodeValues [i], vxId);
```
Comment (antares1470) [inline CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java:255]: final

**Review by antares1470 — CHANGES_REQUESTED** [review]

File: CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java (hunk 2 above)
Comment (CrucisGamma) [reply CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java]: Made the change!

File: CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java (hunk 3 above)
Comment (CrucisGamma) [reply CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java]: Made the change!

File: CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java (hunk 4 above)
Comment (CrucisGamma) [reply CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java]: Made the change!

File: CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java (hunk 5 above)
Comment (CrucisGamma) [reply CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java]: Made the change!

File: CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java (hunk 6 above)
Comment (CrucisGamma) [reply CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/HashmodAction.java]: Made the change!

**Review by antares1470 — CHANGES_REQUESTED** [review]
Forgot to mention this last time but it would be good to add a whatsnew entry for this

**Comment by CrucisGamma** [comment]
Added the what's new entry!

**Review by antares1470 — APPROVED** [review]

**Review by arcturus2 — APPROVED** [review]

**Pull request merged**